### PR TITLE
Add support for `Satisfy` on `ReferenceTypeAssertions`

### LINKED ISSUE #2516: Allow `Should().Satisfy()` on objects and not only collections
### Background and motivation

When I'm asserting many properties on multiple instances, I often find it hard to keep track of the context of what I'm asserting.
Like with collections, we have a `collection.Should().AllSatisfy(item => { /* do FluentAssertions on item here */ }`, but I feel like this is missing for single instances.

The closest thing I've found is `instance.Should().Match(/* something that returns true or false */)`, which is not really fluent as I see it.
The alternative is to do object comparison, where I need to construct the expected object before and then do `instance.Should().BeEquivalentTo(expected)`.

What I'm suggesting is to allow scoped assertions on a single instance, like this:
```cs
var person = new Person { Name = "Peter", Age = 30 };

person.Should().Satisfy(p => 
{
    p.Name.Should().StartWith("P");
    p.Age.Should().BeGreaterThan(25);
});
```

Providing this, also allows me to extract generic assertions as methods, which can be reused.

I am aware that the above is equivalent to this:
```cs
var person = new Person { Name = "Peter", Age = 30 };

person.Name.Should().StartWith("P");
person.Age.Should().BeGreaterThan(25);
```

However, if I have many assertions where the variable names are long, it becomes quite a lot of noise in the tests.
For instance:
```cs
Person someInstanceOfPersonComingFormACallbackFromNSubstitute;

someInstanceOfPersonComingFormACallbackFromNSubstitute.Name.Should().StartWith("P");
someInstanceOfPersonComingFormACallbackFromNSubstitute.Age.Should().BeGreaterThan(25);
```

This would be, IMHO, more readable like this:
```cs
Person someInstanceOfPersonComingFormACallbackFromNSubstitute;

someInstanceOfPersonComingFormACallbackFromNSubstitute.Should().Satisfy(p => 
{
    p.Name.Should().StartWith("P");
    p.Age.Should().BeGreaterThan(25);
});
```

I am well aware that there might be a good reason for this to not exist and that it's possible to create a simple extension method like this:
```cs
public static void ShouldSatisfy<T>(this  T actual, Action<T> assert) where T : class
{
    assert(actual);
}
```

I also understand that this approach could lead to confusion since the `Satisfy()` method would have to exist on all types being asserted.

It might also be the case that my tests are too complex, however, it's an integration test, so I have a lot to assert, which makes the need clear to me.

### Alternative Concerns

_No response_

### Are you willing help with a pull-request?

Yes, please assign this issue to me.

### PR DESCRIPTION
**Pull request by siewers**
Added support for `Satisfy` on `ReferenceTypeAssertions` allowing to use element inspectors as an alternative to using predicates on Match.

Example:
```cs
var myDto = new Dto { Name = "Some Name" };

myDto.Should().Satisfy<Dto>(dto => dto.Name.Should().Be("Some Name"));
```

Fixed #2516

## IMPORTANT 

* [x] If the PR touches the public API, the changes have been approved in a separate issue with the "api-approved" label.
* [x] The code complies with the [Coding Guidelines for C#](https://www.csharpcodingguidelines.com/).
* [x] The changes are covered by unit tests which follow the Arrange-Act-Assert syntax and the naming conventions such as is used [in these tests](../tree/develop/Tests/FluentAssertions.Equivalency.Specs/MemberMatchingSpecs.cs#L51-L430).
* [x] If the PR adds a feature or fixes a bug, please update [the release notes](../tree/develop/docs/_pages/releases.md) with a functional description that explains what the change means to consumers of this library, which are published on the [website](https://fluentassertions.com/releases).
* [x] If the PR changes the public API the changes needs to be included by running [AcceptApiChanges.ps1](../tree/develop/AcceptApiChanges.ps1) or [AcceptApiChanges.sh](../tree/develop/AcceptApiChanges.sh).
* [x] If the PR affects [the documentation](../tree/develop/docs/_pages), please include your changes in this pull request so the documentation will appear on the [website](https://www.fluentassertions.com/introduction).
    * [x] Please also run `./build.sh --target spellcheck` or `.\build.ps1 --target spellcheck` before pushing and check the good outcome

### CODE REVIEW

File: Tests/FluentAssertions.Specs/Primitives/ReferenceTypeAssertionsSpecs.Satisfy.cs
Comment (ITaluone) [inline Tests/FluentAssertions.Specs/Primitives/ReferenceTypeAssertionsSpecs.Satisfy.cs:16]: We usually do not write this `Action x....; x.Should.NotThrow()` anymore, because if it fails, it fails anyway

**Comment by ITaluone** [comment]
Missing release notes?

File: Tests/FluentAssertions.Specs/Primitives/ReferenceTypeAssertionsSpecs.Satisfy.cs
Comment (siewers) [reply Tests/FluentAssertions.Specs/Primitives/ReferenceTypeAssertionsSpecs.Satisfy.cs:16]: Okay?
It just combines the act and assert, which I find a bit awkward. The reason it fails is also unclear from the test since it might fail for any number of reasons.
However, I'm fine doing whatever you find best and there is no consistency in how the tests are written, so it's difficult to get right. I tried to follow the pattern of the tests for the class I am modifying, but even in that class, there is no consistency.

@dennisdoomen Is there a preferred way of writing tests? I can't find any guidelines, so I'd like some "official" answer to this.
The same goes for the usage of nested classes, although I can that pattern being used in several places, I just missed that.

**Comment by siewers** [comment]
I didn't add it to the release notes yet as I'm unsure what version this will be in.

File: Tests/FluentAssertions.Specs/Primitives/ReferenceTypeAssertionsSpecs.Satisfy.cs
Comment (dennisdoomen) [reply Tests/FluentAssertions.Specs/Primitives/ReferenceTypeAssertionsSpecs.Satisfy.cs]: For the test that are not supposed to throw, we do not use the `NotThrow` anymore.

Comment (siewers) [reply Tests/FluentAssertions.Specs/Primitives/ReferenceTypeAssertionsSpecs.Satisfy.cs]: I've merged the Act / Assert for non-throwing assertions

**Comment by ITaluone** [comment]
> I didn't add it to the release notes yet as I'm unsure what version this will be in.

I think while on develop for the next version it doesn't really matter

**Comment by siewers** [comment]
@ITaluone Alright, I added it the the most recent section. I guess it will be aligned when the final release goes out, however, it might be a bit confusing for people using the Alpha versions to see the release notes not reflecting the latest version.

**Comment by siewers** [comment]
@dennisdoomen @ITaluone I believe I've addressed all comments now.

File: Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs
Comment (dennisdoomen) [inline Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs:421]: 🤔 I think that this can be phrased better. We're providing a method that is supposed to be used to execute one or more assertions on an object. And if you run multiple assertions that fail, they will be reported as one failure. We need to capture that in the name.

File: Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs
Comment (dennisdoomen) [inline Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs:423]: 🤔 Same here. We're not satisfying an inspector but executing one or more assertions.

File: Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs
Comment (dennisdoomen) [inline Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs:444]: 🤔 Do we even need to do this? What if you just relied on the assertion scope to throw a combined message? What kind of information would you be missing?

File: docs/_pages/basicassertions.md
Comment (dennisdoomen) [inline docs/_pages/basicassertions.md:87]: ⛏️ The layout is a bit messed up here.

File: docs/_pages/releases.md
Comment (dennisdoomen) [inline docs/_pages/releases.md:18]:
```suggestion
* Introduced a new `Satisfy` method available to all reference types to allow for nested assertions - [#2597](https://github.com/fluentassertions/fluentassertions/pull/2597)
```

File: Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs
Comment (siewers) [reply Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs:444]: I don't know, I'll test it to see if it's necessary 👍

File: Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs
Comment (siewers) [reply Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs:421]: How about:
>Asserts that all element inspectors for the specific instance are satisfied
I'll see if it can be aligned more closely to what is written elsewhere (this one is similar to what is written for `Match`).

Comment (dennisdoomen) [reply Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs:421]: I doubt whether the term "element inspectors" will resonate well with people (even though we've used it before). 

What about 

```csharp
    /// <summary>
    /// Allows combining one or more assertions using the other assertion methods that this library offers on an instance of <typeparamref name="T"/>.
    /// </summary>
    /// <remarks>
    /// If multiple assertions executed by the <paramref name="assertion"/> fail, they will be raised as a single failure.
    /// </remarks>
    /// <param name="assertion">The element inspector which must be satisfied by the <typeparamref name="TSubject" />.</param>
    /// <returns>An <see cref="AndConstraint{T}" /> which can be used to chain assertions.</returns>
    /// <exception cref="ArgumentNullException"><paramref name="assertion"/> is <see langword="null"/>.</exception>
    public AndConstraint<TAssertions> Satisfy<T>(Action<T> assertion)
```

File: Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs
Comment (dennisdoomen) [reply Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs:444]: And?

**Comment by dennisdoomen** [comment]
> I believe I've addressed all comments now.

Apparently not.

File: Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs
Comment (siewers) [reply Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs:444]: Sorry, I haven't had time to do it yet (I'm working on this in my company's "personal time", which is only every second week).
I will look into it this evening.

Comment (siewers) [reply Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs:444]: I tested the changes you suggested, and it looks like we will be losing the outer object context. There is also no indentation on the nested assertions, which I think is useful.

I don't know what is preferred though, however, this is how it's done in `Match`, and I believe the result should be the same or similar, no?

Using a single assertion scope:
```
Expected person.Name to be the same string, but they differ at index 1:
    ↓ (actual)
  "Buford Howard Tannen"
  "Biff Tannen"
    ↑ (expected).
Expected person.Age to be 48, but found 87 (difference of 39).
Expected address.City to be "Hill Valley, San Diego County, California" with a length of 41, but "Hill Valley" has a length of 11, differs near "y" (index 10).
```

Using `Discard`:
```
Expected complexDto to match inspector, but the inspector was not satisfied:
	Expected dto.Person to match inspector, but the inspector was not satisfied:
		Expected person.Name to be the same string, but they differ at index 1:
		    ↓ (actual)
		  "Buford Howard Tannen"
		  "Biff Tannen"
		    ↑ (expected).
		Expected person.Age to be 48, but found 87 (difference of 39).
	Expected dto.Address to match inspector, but the inspector was not satisfied:
		Expected address.City to be "Hill Valley, San Diego County, California" with a length of 41, but "Hill Valley" has a length of 11, differs near "y" (index 10).
```

File: Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs
Comment (siewers) [reply Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs]: Sounds good to me :+1:

File: docs/_pages/basicassertions.md
Comment (siewers) [reply docs/_pages/basicassertions.md:88]: Is this better?

File: Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs
Comment (dennisdoomen) [reply Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs]: You have not addressed this one either?

File: Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs
Comment (dennisdoomen) [reply Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs:447]: When I rely solely on the assertion scope, I get this

```
Expected dto.Person to be the same string, but they differ at index 1:
    ↓ (actual)
  "Buford Howard Tannen"
  "Biff Tannen"
    ↑ (expected).
Expected dto.Person to be 48, but found 87 (difference of 39).
Expected dto.Address to be "Hill Valley, San Diego County, California" with a length of 41, but "Hill Valley" has a length of 11, differs near "y" (index 10).
```

I personally feel like this is much more readable without all the noisy mentioning of the inspector (which is the wrong term anyway), don't you agree?

Comment (siewers) [reply Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs:447]: Something is different in your case. It's not the Person or Address properties of the dto that failed, it's the person.Name and address.City. I think the assertion is presenting the wrong thing.
However, if it's at least showing the correct property in context of the parent, I believe either case is fine. Again, this is based on the implementation Match, so I'm not really trying to make any decisions, I'm just attempting to following an existing pattern 🙂

File: Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs
Comment (siewers) [reply Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs]: I must be confused. I thought both comments were addressed by your suggestion? Honestly, I'm not used to working with GitHub code reviews, so I'm sorry if I'm a bit slow (I'm using Azure DevOps daily, so it's a very different way of working 🙂)

**Comment by ITaluone** [comment]
@siewers Could you rebase to current develop and force push again?

File: Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs
Comment (dennisdoomen) [reply Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs:447]: Just so we talk about the same thing. Which test did you use to reproduce this scenario?

Comment (siewers) [reply Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs:447]: The `Complex_typed_object_not_satisfying_inspector_throws` test.

File: Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs
Comment (IT-VBFK) [inline Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs:429]: I think `assertion` is a bit misleading here 🤔

Comment (dennisdoomen) [reply Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs:429]: You're supposed to use that action to build a composite assertion, so I think it's a suitable term.

File: Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs
Comment (siewers) [reply Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs:447]: What's the conclusion on this one?

Comment (dennisdoomen) [reply Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs:447]: I need to investigate this one a bit better, but haven't had any free time this week. Will try to take a look this weekend.

Comment (dennisdoomen) [reply Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs:447]: I tried again and that made me realize that I don't like the parts about `Expected complexDto to match inspector, but the inspector was not satisfied:`. My suggestion is to change the implementation to 

```csharp
        using (new AssertionScope(CallerIdentifier.DetermineCallerIdentity()))
        {
            assertion((T)Subject);
        }

        return new AndConstraint<TAssertions>((TAssertions)this);
```

This will make your test result in 

```
Expected dto.Person to be the same string, but they differ at index 1:
    ↓ (actual)
  "Buford Howard Tannen"
  "Biff Tannen"
    ↑ (expected).
Expected dto.Person to be 48, but found 87 (difference of 39).
Expected dto.Address to be "Hill Valley, San Diego County, California" with a length of 41, but "Hill Valley" has a length of 11, differs near "y" (index 10).
```

I'll create a separate PR to make sure that nested calls to `AssertionScope(string)` will chain the context you provide so that you'll get something like 

```
Expected complexDto/dto.Person to be the same string, but they differ at index 1:
    ↓ (actual)
  "Buford Howard Tannen"
  "Biff Tannen"
    ↑ (expected).
Expected complexDto/dto.Person to be 48, but found 87 (difference of 39).
Expected complexDto/dto.Address to be "Hill Valley, San Diego County, California" with a length of 41, but "Hill Valley" has a length of 11, differs near "y" (index 10).
```

**Comment by dennisdoomen** [comment]
Did you close it on purpose?

**Comment by siewers** [comment]
Wow, no, I didn't... I don't know what happened.

File: Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs
Comment (dennisdoomen) [reply Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs:447]: Created #2607

Comment (ITaluone) [reply Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs:447]:
```
Expected complexDto/dto.Person to be the same string, but they differ at index 1:
    ↓ (actual)
  "Buford Howard Tannen"
  "Biff Tannen"
    ↑ (expected).
Expected complexDto/dto.Person to be 48, but found 87 (difference of 39).
```

I think what @siewers meant was, that this message presenting the wrong "objects" on which the assertion fails.
Both failing on `Person` but the first mentioning a `string` which doesn't match, and the latter suddenly mentioning (for the same property) an `int` which is failing.

Comment (dennisdoomen) [reply Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs:447]: When you rebase this PR on `develop` (and thereby include #2607), you'll get this:

```
Expected complexDto to match inspector, but the inspector was not satisfied:
	Expected dto.Person to match inspector, but the inspector was not satisfied:
		Expected person.Name to be the same string, but they differ at index 1:
		    ↓ (actual)
		  "Buford Howard Tannen"
		  "Biff Tannen"
		    ↑ (expected).
		Expected person.Age to be 48, but found 87 (difference of 39).
	Expected dto.Address to match inspector, but the inspector was not satisfied:
		Expected address.City to be "Hill Valley, San Diego County, California" with a length of 41, but "Hill Valley" has a length of 11, differs near "y" (index 10).
```

It is still a bit noisy because of the superfluous "Expected complexDto to match inspector, but the inspector was not satisfied", but I'm pretty happy with the result.

Comment (siewers) [reply Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs:447]: I've rebased now, but am I supposed to then only the following in `Satisfy`?
```cs
using (new AssertionScope(CallerIdentifier.DetermineCallerIdentity()))
{
    assertion((T)Subject);
}
```
If I do that, I'm getting this assertion message, which isn't the same as you wrote (and doesn't really make sense) :)
```
Expected complexDto/dto.Person to be the same string, but they differ at index 1:
    ↓ (actual)
  "Buford Howard Tannen"
  "Biff Tannen"
    ↑ (expected).
Expected complexDto/dto.Person to be 48, but found 87 (difference of 39).
Expected complexDto/dto.Address to be "Hill Valley, San Diego County, California" with a length of 41, but "Hill Valley" has a length of 11, differs near "y" (index 10).
```

Comment (siewers) [reply Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs:447]: The problem with the message is that it doesn't mention the failed nested property names:
- `complexDto.Person.Name`
- `complexDto.Person.Age`
- `complexDto.Address.City`
If those exact property names aren't mentioned in the message, it's not helpful.

As it looks with the new changes, both the `Name` and `Age` properties gets merged into a single nonsensical error on the same `Person` property.
It's wrong to say that `Person` was expected to be the same string _and_ it was expected to be a certain number :)
Also, `Address` is another "complex" object, but the error message doesn't present the exact property that failed the assertion.

I hope that makes sense :)

Comment (dennisdoomen) [reply Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs:447]: I need to look into that (again). Later today.

File: Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs
Comment (jnyrup) [inline Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs:443]: If `Satisfy` is used inside an `AssertionScope`, then the execution won't stop if `Subject is T` above fails, meaning we would send in a `null` to `action`.

```cs
object subject = null;

Action act = () =>
{
    using var _ = new AssertionScope();
    subject.Should().Satisfy<object>(x => x.Should().NotBeNull());
};
```

What we usually do to guard against this.
```cs
bool success = Execute.Assertion
    .ForCondition(...)
    .FailWith(...);

if (success)
{
    Execute.Assertion
        .ForCondition(...)
        .FailWith(...);
}
```

File: Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs
Comment (dennisdoomen) [reply Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs:447]: Crap, you're right. The call to `new AssertionScope(CallerIdentifier.DetermineCallerIdentity())` doesn't actually help at all. Sorry about that. I thought I could fix that by changing some of the behavior `AssertionScope` uses. But it turns out, it doesn't. So maybe we can rewrite `Expected dto.Person to match inspector, but the inspector was not satisfied:` to be less verbose.

Comment (jnyrup) [reply Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs:447]: Please also note that `CallerIdentifier.DetermineCallerIdentity()` is very expensive to call and is only needed when the test fails.
See #1660 and #1657

File: Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs
Comment (siewers) [reply Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs]: I've added a check for null, although the message is almost duplicated.
I also used `.Then` to chain the calls, but I'm not sure that does the same thing as you've suggested using the result of the first assertion to execute the next.

File: Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs
Comment (siewers) [reply Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs:447]: I don't think I understand exactly what you mean.
Is it about using an assertion scope within `Satisfy` like this:
```cs
personAndAddressDto.Should().Satisfy<PersonAndAddressDto>(dto => {
    dto.Person.Name.Should().NotBeNullOrEmpty();
    using(new AssertionScope())
    {
        dto.Address.Should().Satisfy<AddressDto>(address => address.City.Should().NotBeNullOrEmpty());
    }
}
```

After I added the null-check, it seems to work as I would expect.

I'm not sure why anyone would add a nested assertion scope though, but of course it needs to be handled correctly.

File: Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs
Comment (dennisdoomen) [reply Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs:451]: > Please also note that `CallerIdentifier.DetermineCallerIdentity()` is very expensive to call and is only needed when the test fails.

True, but it's no longer relevant. My suggestion doesn't work, so the original implementation is still the best.

File: Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs
Comment (dennisdoomen) [reply Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs:447]: What @jnyrup means is that if you're new `Satisfy` is wrapped in an `AssertionScope` (just like you do yourself), the `FailWith` call with _not_ throw an exception and return `false` instead.

File: Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs
Comment (siewers) [reply Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs:429]: It might, but I don't know how to connect that to the nested assertions or how the resulting assertion message should look.
In practice, the nested assertions can have the `because`, but the parent would probably always have the message "because the assertions should be satisfied" or something like that.

File: Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs
Comment (siewers) [reply Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs:447]: I'm so sorry for the late reply. I will try to support that.

Comment (siewers) [reply Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs:447]: Does this test verify the expected behavior (it passes with the current implementation)?

```cs
[Fact]
public void Using_assertion_scope_with_null_subject()
{
    // Arrange
    object subject = null;

    // Act
    Action act = () =>
    {
        using (new AssertionScope())
        {
            subject.Should().Satisfy<object>(x => x.Should().NotBeNull());
        }
    };

    // Assert
    act.Should().Throw<XunitException>()
        .WithMessage("Expected subject to be assignable to System.Object, but found <null>.");
}
```

If this is not what you mean, I need some help understanding the scenario you have in mind.

**Review by dennisdoomen — APPROVED** [review]

**Comment by siewers** [comment]
@dennisdoomen I see you approved this PR, but it's blocked by a couple of checks. Is there anything further on my part I need to take care of?

**Comment by IT-VBFK** [comment]
Rebase to current develop and force push

**Comment by ITaluone** [comment]
@siewers To fix the current problems rebase again :)

**Comment by siewers** [comment]
I've rebased, and all checks are passing. Can we merge this?

**Comment by dennisdoomen** [comment]
> I've rebased, and all checks are passing. Can we merge this?

Waiting for a review from @jnyrup

File: Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs
Comment (jnyrup) [reply Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs:447]: Your `Using_assertion_scope_with_null_subject` reveals that using `Subject.GetType()` throws an NRE and needs to be `Subject?.GetType()`

If we also change `Typed_object_satisfied_against_incorrect_type_throws` to have 

```cs
Action act = () =>
{
    using var _ = new AssertionScope();
    personDto.Should().Satisfy<AddressDto>(dto => dto.Should().NotBeNull());
};
```

that reveals that `assertion((T)Subject);` throws `System.InvalidCastException: 'Unable to cast object of type 'PersonDto' to type 'AddressDto'.'`

We usually handle this by return the return value of the assertion chain and check if it was successful.

```cs
bool success = Execute.Assertion
    .ForCondition(Subject is not null)
    .WithDefaultIdentifier(Identifier)
    .FailWith("Expected {context:object} to be assignable to {0}{reason}, but found <null>.", typeof(T))
    .Then
    .ForCondition(Subject is T)
    .WithDefaultIdentifier(Identifier)
    .FailWith("Expected {context:object} to be assignable to {0}{reason}, but {1} is not.", typeof(T), Subject?.GetType());

if (success)
{
    string[] failuresFromInspector;

    using (var assertionScope = new AssertionScope())
    {
        assertion((T)Subject);
        failuresFromInspector = assertionScope.Discard();
    }

    if (failuresFromInspector.Length > 0)
    {
        string failureMessage = Environment.NewLine
            + string.Join(Environment.NewLine, failuresFromInspector.Select(x => x.IndentLines()));

        Execute.Assertion
            .WithDefaultIdentifier(Identifier)
            .WithExpectation("Expected {context:object} to match inspector, but the inspector was not satisfied:", Subject)
            .FailWithPreFormatted(failureMessage);
    }
}

return new AndConstraint<TAssertions>((TAssertions)this);
```

Comment (siewers) [reply Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs]: I am unable to reproduce the exception with the suggested test.
Have you tested whether it throws an NRE?

I understand that the combination of `Execute.Assertion` and `AssertionScope` should behave in some specific way, but I don't know how that is.

In any case, I added the precondition check and the test, as suggested, even though there is no difference, regardless of what I do.

Could you please test it? I feel I'm wasting too much of your time because I don't understand the underlying problems you are trying to explain.

Comment (jnyrup) [reply Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs]: Run `Using_assertion_scope_with_null_subject` with a debugger attached and for a moment remove
https://github.com/fluentassertions/fluentassertions/blob/8656682fe60df5ebff2cd6cc7a0d3c140e96786b/Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs#L15

When using an `AssertionScope` failing calls to `ForCondition` doesn't immediately cause the entire assertion to stop.
This is usually used for doing multiple assertions on different subjects and get all failures in the output.
E.g. if both of these assertions fail we get two failure messages instead of only the one from the first assertion.
```cs
using (new AssertionScope())
{
    a.Should().BeFoo();
    b.Should().BeBar();
}
```

As `AssertionScope`s continues the execution, we also internally in FluentAssertions need to guard against the execution continuing.

So when wrapping the call to `Satisfy` in an `AssertionScope` the assertion continues executing after `Subject is not null` fails, which leads to the NRE when calling `Subject!.GetType()` on a null `Subject`.
image

The fix here is to change `Subject!.GetType()` to `Subject?.GetType()`

Your test `Using_assertion_scope_with_subject_satisfied_against_incorrect_type_throws` correctly exercises the case where  we continue after `Subject is T` fails and `if (success)` guards against casting `Subject` to `T`.
image

Comment (siewers) [reply Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs]: Ah, okay, now I'm starting to understand the problem.
The tests are not throwing the NREs, but internally, Fluent Assertions are not executing the assertion because it fails to handle the second precondition.

I mistakenly assumed `.Then` only got executed if the previous assertion succeeded.

Would it be better to split the checks instead of using `.Then`?

**Review by jnyrup — APPROVED** [review]

File: Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs
Comment (jnyrup) [reply Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs]: Your fix with replacing `!` with `?` is exactly what I would do.

**Pull request merged**
